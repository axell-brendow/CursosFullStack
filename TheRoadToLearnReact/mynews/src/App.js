import React from 'react';
import logo from './logo.svg';
import './App.css';

const DEFAULT_QUERY = 'redux';

const PATH_BASE = 'https://hn.algolia.com/api/v1';
const PATH_SEARCH = '/search';
const PARAM_SEARCH = 'query=';

let url = `${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${DEFAULT_QUERY}`;

const list = [
  {
    title: 'React',
    url: 'https://reactjs.org/',
    author: 'Jordan Walke',
    num_comments: 3,
    points: 4,
    objectID: 0,
  },
  {
    title: 'Redux',
    url: 'https://redux.js.org/',
    author: 'Dan Abramov, Andrew Clark',
    num_comments: 2,
    points: 5,
    objectID: 1,
  },
];

class Button extends React.Component {
  render() {
    const { onClick, className = '', children, } = this.props;

    return (
      <button
        onClick={onClick}
        className={className}
        type="button"
      >
        {children}
      </button>
    );
  }
}

function Search({ value, onChange, children })
{
  return (
    <form>
      { children }
      <input type="text" value={value} onChange={onChange}></input>
    </form>
  );
}

function filterByTerm(searchTerm)
{
  return (item) => item.title.toLocaleLowerCase().includes( searchTerm.toLocaleLowerCase() );
}

function Table({ list, pattern, onDismiss })
{
  const largeColumn = { width: '40%' };
  const midColumn = { width: '30%' };
  const smallColumn = { width: '10%' };

  return (
    <table className="table">
      <thead>
        <tr>
          <th>Title</th>
          <th>Author</th>
          <th>Number of comments</th>
          <th>Points</th>
          <th>Remove</th>
        </tr>
      </thead>
      <tbody>
      {
        list.filter(
          filterByTerm(pattern)
        )
        .map(
          (item) =>
          <tr key={ item.objectID }>
            <td style={ largeColumn }>
              <a href={ item.url }>{ item.title }</a>
            </td>
            <td style={ midColumn }>{ item.author }</td>
            <td style={ smallColumn }>{ item.num_comments }</td>
            <td style={ smallColumn }>{ item.points }</td>
            <td style={ smallColumn }>
              <Button onClick={ () => onDismiss(item.objectID) }
                className="button-inline">
                Dismiss
              </Button>
            </td>
          </tr>
        )
      }
      </tbody>
    </table>
  );
}

class App extends React.Component
{
  constructor(props)
  {
    super(props);

    this.state = {
      result: null,
      searchTerm: DEFAULT_QUERY
    };

    this.setSearchTopStories = this.setSearchTopStories.bind(this);
    this.onDismiss = this.onDismiss.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
  }

  setSearchTopStories(result)
  {
    this.setState({ result });
  }

  componentDidMount()
  {
    const { searchTerm } = this.state;
    
    url = `${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}`;

    fetch(url).then((response) => response.json())
      .then((json) => this.setSearchTopStories(json))
      .catch(error => error);
  }

  onDismiss(id)
  {
    const updatedList = this.state.result.hits.filter(
      (item) => item.objectID != id
    );

    this.setState(
      {
        // No create-react-app, o operador de spread ... cria uma c√≥pia das propriedades do objeto
        result: { ...this.state.result, hits: updatedList }
      }
    );
  }

  onSearchChange(event)
  {
    this.setState({ searchTerm: event.target.value });
  }

  render()
  {
    const {result, searchTerm} = this.state;
    let element = null;

    if (result)
    {
      element = (
        <div className="page">
          <div className="interactions"> 
            <Search value={searchTerm} onChange={this.onSearchChange}>
              Search
            </Search>
          </div>
            <Table list={result.hits} pattern={searchTerm} onDismiss={this.onDismiss}></Table>
        </div>
      );
    }

    return element;
  }
}

export default App;
