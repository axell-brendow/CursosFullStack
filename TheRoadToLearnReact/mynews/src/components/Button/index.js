import React from 'react';

// O React vem com um checador de tipos embutido: PropTypes.
import PropTypes from 'prop-types';

export default class Button extends React.Component {
  render() {
    const { onClick, className, type = 'button', children } = this.props;
    
    return (
      <button
        onClick={onClick}
        className={className}
        type={type}
      >
        {children}
      </button>
    );
  }
}

Button.propTypes = {
  onClick: PropTypes.func.isRequired,
  className: PropTypes.string,
  children: PropTypes.node.isRequired
};

Button.defaultProps = {
  className: ''
};