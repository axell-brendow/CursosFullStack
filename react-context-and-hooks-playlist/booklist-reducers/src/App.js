import React from 'react';
import BookContextProvider from './contexts/BookContext';
import Navbar from './components/Navbar';
import BookList from './components/BookList';
import BookForm from './components/BookForm';
import ReducerTest from './components/ReducerTest';

function App() {
  return (
    <div className="App">
      <BookContextProvider>
        <Navbar/>
        <BookList/>
        <BookForm/>
        <ReducerTest/>
      </BookContextProvider>
    </div>
  );
}

export default App;
