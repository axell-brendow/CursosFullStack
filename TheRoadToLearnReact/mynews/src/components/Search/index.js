import React from 'react';
import Button from '../Button';

export default function Search({ value, onChange, onSubmit, children })
{
  return (
    <form onSubmit={onSubmit}>
      <input type="text" value={value} onChange={onChange}></input>
      <Button type="submit">
        { children }
      </Button>
    </form>
  );
}
