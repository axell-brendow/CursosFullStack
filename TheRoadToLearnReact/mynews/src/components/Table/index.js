import React from 'react';
import Button from '../Button';

function filterByTerm(searchTerm)
{
  return (item) =>
  {
    let include = false;

    if (searchTerm && item && item.title)
    {
      include = item.title.toLocaleLowerCase()
        .includes( searchTerm.toLocaleLowerCase() );
    }

    return include;
  };
}

export default function Table({ list, pattern, onDismiss })
{
  const largeColumn = { width: '40%' };
  const midColumn = { width: '30%' };
  const smallColumn = { width: '10%' };

  return (
    <table className="table">
      <thead>
        <tr>
          <th>Title</th>
          <th>Author</th>
          <th>Number of comments</th>
          <th>Points</th>
          <th>Remove</th>
        </tr>
      </thead>
      <tbody>
      {
        list.filter(
          filterByTerm(pattern)
        )
        .map(
          (item) =>
          <tr key={ item.objectID }>
            <td style={ largeColumn }>
              <a href={ item.url }>{ item.title }</a>
            </td>
            <td style={ midColumn }>{ item.author }</td>
            <td style={ smallColumn }>{ item.num_comments }</td>
            <td style={ smallColumn }>{ item.points }</td>
            <td style={ smallColumn }>
              <Button onClick={ () => onDismiss(item.objectID) }
                className="button-inline">
                Dismiss
              </Button>
            </td>
          </tr>
        )
      }
      </tbody>
    </table>
  );
}
