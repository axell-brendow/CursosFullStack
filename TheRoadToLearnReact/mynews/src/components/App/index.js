import React from 'react';
import axios from 'axios';

// O React vem com um checador de tipos embutido: PropTypes.
import PropTypes from 'prop-types';

// import React from '../../../node_modules/react';
// import axios from '../../../node_modules/axios';
import Button from '../Button';
import Search from '../Search';
import Table from '../Table';
import logo from '../../logo.svg';
import './index.css';

import {
  DEFAULT_QUERY,
  DEFAULT_HITS_PER_PAGE,
  PATH_BASE,
  PATH_SEARCH,
  PARAM_SEARCH,
  PARAM_PAGE,
  PARAM_HPP,
} from '../../constants';
import {Link} from "react-router-dom";

let url = `${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${DEFAULT_QUERY}&${PARAM_PAGE}&${PARAM_HPP}${DEFAULT_HITS_PER_PAGE}`;

class App extends React.Component
{
  constructor(props)
  {
    super(props);

    this.state = {
      results: null,
      searchKey: '',
      searchTerm: DEFAULT_QUERY
    };

    this.needsToSearchTopStories = this.needsToSearchTopStories.bind(this);
    this.setSearchTopStories = this.setSearchTopStories.bind(this);
    this.fetchSearchTopStories = this.fetchSearchTopStories.bind(this);
    this.onDismiss = this.onDismiss.bind(this);
    this.onSearchChange = this.onSearchChange.bind(this);
    this.onSearchSubmit = this.onSearchSubmit.bind(this);
  }

  needsToSearchTopStories(searchTerm)
  {
    return !this.state.results[searchTerm];
  }
  
  setSearchTopStories(result)
  {
    const { hits, page } = result;
    const { searchKey, results } = this.state;
    
    const oldHits = (results && results[searchKey]) ? results[searchKey].hits : [];
    const updatedHits = [ ...oldHits, ...hits ];
    
    this.setState({
      results: {
        ...results,
        [searchKey]: { hits: updatedHits, page }
      }
    });
  }

  fetchSearchTopStories(searchTerm, page = 0)
  {
    url = `${PATH_BASE}${PATH_SEARCH}?${PARAM_SEARCH}${searchTerm}&${PARAM_PAGE}${page}&${PARAM_HPP}${DEFAULT_HITS_PER_PAGE}`;

    axios(url).then( (result) => this.setSearchTopStories(result.data) )
      .catch( (error) => this.setState( { error } ) );
  }
  
  componentDidMount()
  {
    const { searchTerm } = this.state;
    
    this.setState({ searchKey: searchTerm });
    this.fetchSearchTopStories(searchTerm);
  }

  onDismiss(id)
  {
    const { results, searchKey } = this.state;
    const { hits, page } = results[searchKey];
    
    const updatedList = hits.filter(
      (item) => item.objectID !== id
    );

    this.setState(
      {
        // No create-react-app, o operador de spread ... cria uma c√≥pia das propriedades do objeto
        results: {
          ...results,
          [searchKey]: { hits: updatedList, page }
        }
      }
    );
  }

  onSearchChange(event)
  {
    this.setState({ searchTerm: event.target.value });
  }

  onSearchSubmit(event)
  {
    const { searchTerm } = this.state;
    
    this.setState({ searchKey: searchTerm });
    
    if (this.needsToSearchTopStories(searchTerm))
    {
      this.fetchSearchTopStories(searchTerm);
    }
    
    event.preventDefault();
  }
  
  render()
  {
    const { searchTerm, results, searchKey, error } = this.state;
    const resultsForCurrentKey = results && results[searchKey];
    
    const page = (
      resultsForCurrentKey && results[searchKey].page
    ) || 0;
    
    const list = (
      resultsForCurrentKey && results[searchKey].hits
    ) || [];
    
    let element = null;

    if (resultsForCurrentKey)
    {
      element = (
        <div className="page">
          <div className="interactions"> 
            <Search value={searchTerm} onChange={this.onSearchChange} onSubmit={this.onSearchSubmit}>
              Search
            </Search>
            <Link to={"/about"}>About</Link>
          </div>
          
          {
            error ?
            <p>Something went wrong.</p> :
            <Table list={list} pattern={searchTerm} onDismiss={this.onDismiss}/>
          }
          
          <div className="interactions"> 
            <Button onClick={() => this.fetchSearchTopStories(searchKey, page + 1)}>
              more
            </Button>
          </div>
        </div>
      );
    }

    return element;
  }
}

export default App;

export {
  Button,
  Search,
  Table
};