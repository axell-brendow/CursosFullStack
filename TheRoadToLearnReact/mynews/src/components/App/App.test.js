import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import App from '.';
import Search from '../Search';
import Button from '../Button';
import Table from '../Table';
import { exportAllDeclaration } from '@babel/types';

describe('App',
  () =>
  {
    it('renders without crashing', // teste simples do componente
      () => {
        const div = document.createElement('div');
        ReactDOM.render(<App />, div);
        ReactDOM.unmountComponentAtNode(div);
      }
    );

    test('has a valid snapshot', // teste do test-renderer com snapshot
      () =>
      {
        const component = renderer.create( <App/> );
        const tree = component.toJSON();

        expect(tree).toMatchSnapshot();
      }
    );
  }
);

describe('Search',
  () =>
  {
    it('renders without crashing', // teste simples do componente
      () => {
        const div = document.createElement('div');
        ReactDOM.render(<Search>Search</Search>, div);
        ReactDOM.unmountComponentAtNode(div);
      }
    );

    test('has a valid snapshot', // teste do test-renderer com snapshot
      () =>
      {
        const component = renderer.create( <Search>Search</Search> );
        const tree = component.toJSON();

        expect(tree).toMatchSnapshot();
      }
    );
  }
);

describe('Button',
  () =>
  {
    it('renders without crashing', // teste simples do componente
      () => {
        const div = document.createElement('div');
        ReactDOM.render(<Button>Give Me More</Button>, div);
        ReactDOM.unmountComponentAtNode(div);
      }
    );

    test('has a valid snapshot', // teste do test-renderer com snapshot
      () =>
      {
        const component = renderer.create( <Button>Give Me More</Button> );
        const tree = component.toJSON();

        expect(tree).toMatchSnapshot();
      }
    );
  }
);

describe('Table',
  () =>
  {
    const list = [
      {
        title: 'React',
        url: 'https://reactjs.org/',
        author: 'Jordan Walke',
        num_comments: 3,
        points: 4,
        objectID: 0,
      },
      {
        title: 'Redux',
        url: 'https://redux.js.org/',
        author: 'Dan Abramov, Andrew Clark',
        num_comments: 2,
        points: 5,
        objectID: 1,
      },
    ];

    it('renders without crashing', // teste simples do componente
      () => {
        const div = document.createElement('div');
        ReactDOM.render(<Table list={list}/>, div);
        ReactDOM.unmountComponentAtNode(div);
      }
    );

    test('has a valid snapshot', // teste do test-renderer com snapshot
      () =>
      {
        const component = renderer.create( <Table list={list}/> );
        const tree = component.toJSON();

        expect(tree).toMatchSnapshot();
      }
    );
  }
);
