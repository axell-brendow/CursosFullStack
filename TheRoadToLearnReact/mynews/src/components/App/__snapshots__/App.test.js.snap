// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App has a valid snapshot 1`] = `null`;

exports[`Button has a valid snapshot 1`] = `
<button
  className=""
  type="button"
>
  Give Me More
</button>
`;

exports[`Search has a valid snapshot 1`] = `
<form>
  <input
    type="text"
  />
  <button
    className=""
    type="button"
  >
    Search
  </button>
</form>
`;

exports[`Table has a valid snapshot 1`] = `
<table
  className="table"
>
  <thead>
    <tr>
      <th>
        Title
      </th>
      <th>
        Author
      </th>
      <th>
        Number of comments
      </th>
      <th>
        Points
      </th>
      <th>
        Remove
      </th>
    </tr>
  </thead>
  <tbody />
</table>
`;
